File: web/src/lib/domain/activityIndices.ts
import type { WeatherData } from './weatherData';

export type ActivityTier = 'EXCELLENT' | 'GOOD' | 'FAIR' | 'POOR';

export interface ActivityIndices {
	running: ActivityTier;
	laundry: ActivityTier;
	uvSafety: ActivityTier;
}

export function computeActivityIndices(data: WeatherData): ActivityIndices {
	const tempC = data.temperature.celsius;
	const humidity = data.humidity;
	const windMs = data.windSpeed.metersPerSecond;
	const precipMm = (data.rain?.mm ?? 0) + (data.snow?.mm ?? 0);
	const uv = data.uvIndex;

	let runningTier: ActivityTier = 'FAIR';
	if (precipMm > 0 || tempC < 0 || tempC > 32 || windMs > 12) {
		runningTier = 'POOR';
	} else if (tempC >= 8 && tempC <= 22 && humidity < 70 && windMs < 7) {
		runningTier = 'EXCELLENT';
	} else if (tempC >= 2 && tempC <= 28 && humidity < 85) {
		runningTier = 'GOOD';
	}

	let laundryTier: ActivityTier = 'FAIR';
	if (precipMm > 0 || humidity > 80) {
		laundryTier = 'POOR';
	} else if (tempC > 18 && humidity < 50 && windMs >= 2) {
		laundryTier = 'EXCELLENT';
	} else if (tempC > 12 && humidity < 65) {
		laundryTier = 'GOOD';
	}

	let uvTier: ActivityTier = 'POOR';
	if (uv < 3) {
		uvTier = 'EXCELLENT';
	} else if (uv < 6) {
		uvTier = 'GOOD';
	} else if (uv < 8) {
		uvTier = 'FAIR';
	}

	const result: ActivityIndices = {
		running: runningTier,
		laundry: laundryTier,
		uvSafety: uvTier
	};
	return result;
}

File: app/src/main/java/orinasa/njarasoa/maripanatokana/domain/model/ActivityIndices.kt
package orinasa.njarasoa.maripanatokana.domain.model

enum class ActivityTier {
    EXCELLENT,
    GOOD,
    FAIR,
    POOR
}

data class ActivityIndices(
    val running: ActivityTier,
    val laundry: ActivityTier,
    val uvSafety: ActivityTier,
) {
    companion object {
        fun fromWeatherData(data: WeatherData): ActivityIndices {
            val tempC = data.temperature.celsius
            val humidity = data.humidity
            val windMs = data.windSpeed.metersPerSecond
            val precipMm = (data.rain?.mm ?: 0.0) + (data.snow?.mm ?: 0.0)
            val uv = data.uvIndex

            val runningTier = if (precipMm > 0.0 || tempC < 0.0 || tempC > 32.0 || windMs > 12.0) {
                ActivityTier.POOR
            } else if (tempC in 8.0..22.0 && humidity < 70 && windMs < 7.0) {
                ActivityTier.EXCELLENT
            } else if (tempC in 2.0..28.0 && humidity < 85) {
                ActivityTier.GOOD
            } else {
                ActivityTier.FAIR
            }

            val laundryTier = if (precipMm > 0.0 || humidity > 80) {
                ActivityTier.POOR
            } else if (tempC > 18.0 && humidity < 50 && windMs >= 2.0) {
                ActivityTier.EXCELLENT
            } else if (tempC > 12.0 && humidity < 65) {
                ActivityTier.GOOD
            } else {
                ActivityTier.FAIR
            }

            val uvTier = if (uv < 3.0) {
                ActivityTier.EXCELLENT
            } else if (uv < 6.0) {
                ActivityTier.GOOD
            } else if (uv < 8.0) {
                ActivityTier.FAIR
            } else {
                ActivityTier.POOR
            }

            return ActivityIndices(
                running = runningTier,
                laundry = laundryTier,
                uvSafety = uvTier,
            )
        }
    }
}
